x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamOutput.java

```diff
@@ -181,18 +182,31 @@ public void writeCachedBlock(Block block) throws IOException {
     }
 
     @Override
-    public Integer fromAttributeCache(Attribute attr) {
+    public boolean writeAttributeCacheHeader(Attribute attribute) throws IOException {
+        if (getTransportVersion().onOrAfter(TransportVersions.ESQL_FIELD_ATTRIBUTE_CACHED_SERIALIZATION)) {
+            Integer cacheId = attributeIdFromCache(attribute);
+            if (cacheId != null) {
+                writeZLong(cacheId);
+                return false;
+            }
+
+            cacheId = cacheAttribute(attribute);
+            writeZLong(-1 - cacheId);
+        }
+        return true;
+    }
+
+    private Integer attributeIdFromCache(Attribute attr) {
         return cachedAttributes.get(attr);
     }
 
-    @Override
-    public Integer addToAttributeCache(Attribute attr) {
+    private int cacheAttribute(Attribute attr) {
         if (cachedAttributes.containsKey(attr)) {
             throw new IllegalArgumentException("Attribute already present in the serialization cache [" + attr + "]");
         }
         int id = cachedAttributes.size();
         if (id >= maxAttributeCacheSize) {
```

Comment:
I made it parametric in the constructor, so that
- I can test it with lower values
- unlikely, but If we decide, we can make it parametric per request

Not sure it's worth it, but it doesn't make things too complicated